src/Ship.cpp

```diff
@@ -1256,10 +1256,23 @@ void Ship::StaticUpdate(const float timeStep)
 				const vector3d vdir = GetVelocity().Normalized();
 				const vector3d pdir = -GetOrient().VectorZ();
 				const double dot = vdir.Dot(pdir);
-				if ((m_stats.free_capacity) && (dot > 0.90) && (speed > 100.0) && (density > 0.3)) {
-					const double rate = speed * density * 0.00000333 * double(m_stats.fuel_scoop_cap);
-					if (Pi::rng.Double() < rate) {
-						LuaEvent::Queue("onShipScoopFuel", this, p);
+				const double speed_times_density = speed * density;
+				/*
+				 * speed = m/s, density = g/cm^3 -> T/m^3, pressure = Pa -> N/m^2 -> kg/(m*s^2)
+				 * m    T      kg         m*kg^2           kg^2
+				 * - * --- * ----- = 1000 ------- = 1000 -------
+				 * s   m^3   m*s^2        m^4*s^3        m^3*s^3
+				 *
+				 * fuel_scoop_cap = area, m^2. rate = kg^2/(m*s^3) = (Pa*kg)/s^2
+				 */
+				const double hydrogen_density = 0.00000333;
+				if ((m_stats.free_capacity) && (dot > 0.90) && speed_times_density > (100.0 * 0.3)) {
+					const double rate = speed_times_density * hydrogen_density * double(m_stats.fuel_scoop_cap);
+					m_hydrogenScoopedAccumulator += rate * Pi::game->GetTimeAccelRate();
```

Comment:
`Pi::game->GetTimeAccelRate()` is not the "full picture" when it comes to making this function update-rate independent - you should multiply by `timeStep` instead, as that is the fraction of simulated time elapsed during the update period. You may need to adjust the rate of fuel scooping, as the function is currently scooping ~60x as fast as it should.

Reply:
> as the function is currently scooping ~60x as fast as it should.

I don't actually see why.
Before PR, scooping rate was compared against [0; 1) random number every tick, and if the rate is bigger, a ton of hydrogen is added to your cargo. I kept this rate unchanged.

Did a quick tests:
Scoop at 100 m/s on Saturn (I'll post save later) for 5 minutes.
Atmospheric pressure: ~51.8 atm

Before scooping: 29 tons
After scooping
Old algorithm, try 1: 257 tons (228t scooped)
Old algorithm, try 2: 238 tons (209t scooped)
Old algorithm, try 3: 259 tons (230t scooped)
New algorithm, without 4394d63d3: 239 tons (210t scooped)
New algorithm, without 4394d63d3, x10 acceleration: 239 tons (210t scooped)
New algorithm, with 4394d63d3: 33 tons (4t scooped)

With changes you requested scooping became ~60x times slower

Reply:
Attaching the savefile

[scooping-test.gz](https://github.com/pioneerspacesim/pioneer/files/12642288/scooping-test.gz)

Reply:
I meant that you may need to increase the rate of scooping to (re-)create the intended flow of fuel. The function is currently scooping at a timestep-invariant rate, which will only function correctly so long as the game is updating at a fixed 60 ticks per second. If any change occurs to the frequency of timestep updates, the rate of fuel will change in a manner that is not consistent with the atmospheric density or speed.

The function should be modified to multiply by `timestep`, which you did in https://github.com/pioneerspacesim/pioneer/commit/4394d63d3a3554a0915d1cade9bc4e81a544668f, *and* the rate adjusted upwards by exactly 60x. I specifically called out that it was scooping 60x as fast as you would expect if you tried to do the math by hand using the existing constants in the code.

Reply:
> If any change occurs to the frequency of timestep updates

Is it intended to change?

Reply:
> Is it intended to change?

Assume yes. There's a proposal on IRC right now for adaptive substepping to improve ships crashing into the ground at high time acceleration. Whether or not we intend to change the timestep however, the correct way to express the algorithm is in terms of the timestep value.